### PR TITLE
Implement subscription edge cases (issue #3)

### PR DESCRIPTION
**Pull request by philipz**
## Summary

Implement subscription edge cases (issue #3) - Handle edge cases in subscription lookup:

- **No active subscription**: Returns HTTP 409 when customer has no active subscription (expired or no subscription)
- **Multiple active subscriptions**: Returns HTTP 500 with `log.error("Multiple active subscriptions for customerId={}", customerId)` - treated as data integrity error

### Changes

1. **UsageService.java**: Modified to use new `findAllActiveSubscriptions` method, added `NoActiveSubscriptionException` and `MultipleSubscriptionsException`
2. **CustomerSubscriptionRepository.java**: Added `findAllActiveSubscriptions` method returning List
3. **UsageController.java**: Added exception handlers for new exceptions (409 and 500)
4. **V1__Create_tables.sql**: Added seed data for test cases:
   - `CUST-004`: No active subscription (expired effective_to)
   - `CUST-005`: Multiple active subscriptions (data integrity error)
5. **UsageServiceTest.java**: Added test cases for edge cases

### Acceptance Criteria

- [x] No active subscription → HTTP 409
- [x] Data integrity error → HTTP 500 + log.error
- [x] `effective_from = today` (for active subscriptions)
- [x] `effective_to = today` and `effective_to = today - 1` (for expired)

@philipz can click here to [continue refining the PR](https://app.all-hands.dev/conversations/5549b6e0-d6b6-42fb-927d-2f6f4947dd5f)

### CODE REVIEW

**Review by philipz — COMMENTED** [review]
## Code Review

🟡 Acceptable - Works but could be cleaner

**CRITICAL ISSUES**

None - implementation is functionally correct per issue #3.


**IMPROVEMENT OPPORTUNITIES**

1. Duplicate query in Repository - findActiveSubscription and findAllActiveSubscriptions have identical queries. Consider removing findActiveSubscription if unused.

2. Data in migration - Seed data in V1 migration is not recommended. Consider separate migration file.

3. Missing index - Query filters on effective_from/effective_to but no index. For large tables:
```sql
CREATE INDEX idx_subscriptions_dates ON customer_subscriptions(customer_id, effective_from, effective_to);
```

**TESTING GAPS**

None - unit tests cover edge cases.

**RISK ASSESSMENT**
- Overall: LOW - No breaking changes to public API

**VERDICT**
✅ Worth merging - Core logic is sound

**KEY INSIGHT**
Consider adding index on subscription date columns.

File: src/main/java/org/tw/token_billing/repository/CustomerSubscriptionRepository.java [hunk 1]
```diff
@@ -20,4 +21,15 @@ Optional<CustomerSubscription> findActiveSubscription(
         @Param("customerId") String customerId,
         @Param("effectiveDate") LocalDate effectiveDate
     );
```
[the remark this answers is not in the document]
Comment (philipz) [reply src/main/java/org/tw/token_billing/repository/CustomerSubscriptionRepository.java]: Done in fb0df99: removed `findActiveSubscription`; the service now uses `findAllActiveSubscriptions` exclusively.

File: src/main/java/org/tw/token_billing/controller/UsageController.java [hunk 2]
```diff
@@ -27,9 +27,15 @@ public ResponseEntity<BillResponse> submitUsage(@Valid @RequestBody UsageRequest
         return ResponseEntity.status(HttpStatus.CREATED).body(response);
     }
 
-    @ExceptionHandler(UsageService.CustomerNotFoundException.class)
-    public ResponseEntity<ErrorResponse> handleCustomerNotFound(UsageService.CustomerNotFoundException ex) {
-        return ResponseEntity.status(HttpStatus.NOT_FOUND)
+    @ExceptionHandler(UsageService.NoActiveSubscriptionException.class)
+    public ResponseEntity<ErrorResponse> handleNoActiveSubscription(UsageService.NoActiveSubscriptionException ex) {
+        return ResponseEntity.status(HttpStatus.CONFLICT)
+            .body(new ErrorResponse(ex.getMessage()));
```
[the remark this answers is not in the document]
Comment (philipz) [reply src/main/java/org/tw/token_billing/controller/UsageController.java]: Fixed in fb0df99. `UsageService` now calls `customerRepository.existsById` first per SRS-F-6 step 4: a non-existent customer throws `CustomerNotFoundException` (404), and only customers that exist but have zero active subscriptions reach the 409 path. Covered by the new `calculateBill_customerDoesNotExist_throwsCustomerNotFound` service test (verifies the subscription repo is never queried).

File: src/main/resources/db/migration/V1__Create_tables.sql [hunk 3]
```diff
@@ -52,10 +52,17 @@ INSERT INTO pricing_plans (id, name, monthly_quota, overage_rate_per_1k) VALUES
...
-    ('c3d4e5f6-a7b8-9012-cdef-123456789012', 'CUST-003', 'PLAN-ENTERPRISE', '2026-01-15');
+INSERT INTO customer_subscriptions (id, customer_id, plan_id, effective_from, effective_to) VALUES
+    ('a1b2c3d4-e5f6-7890-abcd-ef1234567890', 'CUST-001', 'PLAN-STARTER', '2026-01-01', NULL),
+    ('b2c3d4e5-f6a7-8901-bcde-f12345678901', 'CUST-002', 'PLAN-FREE', '2026-02-01', NULL),
+    ('c3d4e5f6-a7b8-9012-cdef-123456789012', 'CUST-003', 'PLAN-ENTERPRISE', '2026-01-15', NULL),
+    -- CUST-004: expired subscription (effective_to = today - 1) -> no active subscription
+    ('d4e5f6a7-a8b9-0123-def0-234567890123', 'CUST-004', 'PLAN-STARTER', '2026-01-01', '2026-05-01'),
+    -- CUST-005: multiple active subscriptions -> data integrity error
+    ('e5f6a7b8-b9c0-1234-ef01-345678901234', 'CUST-005', 'PLAN-STARTER', '2026-01-01', NULL),
+    ('f6a7b8c9-c0d1-2345-f012-456789012345', 'CUST-005', 'PLAN-PRO', '2026-01-01', NULL);
```
[the remark this answers is not in the document]
Comment (philipz) [reply src/main/resources/db/migration/V1__Create_tables.sql]: Fixed in fb0df99 with the suggested dates: subscriptions now use `2020-01-01` (active) and `2020-01-01..2020-01-02` for CUST-004 (expired). Fixtures no longer depend on system-clock comparisons crossing 2026-05-01.
